Engine/media/audio/audio.cpp

```diff
@@ -39,8 +39,53 @@
 #include "main/game_run.h"
 
 using namespace AGS::Common;
+using namespace AGS::Engine;
+
+//-----------------------
+//sound channel management; all access goes through here, which can't be done without a lock
+
+static std::array<SOUNDCLIP *,MAX_SOUND_CHANNELS+1> _channels;
+AGS::Engine::Mutex AudioChannelsLock::s_mutex;
+
+SOUNDCLIP *AudioChannelsLock::GetChannel(int index)
+{
+    return _channels[index];
+}
+
+SOUNDCLIP *AudioChannelsLock::GetChannelIfPlaying(int index)
+{
+    auto *ch = _channels[index];
+    return (ch != nullptr && ch->is_playing()) ? ch : nullptr;
+}
+
+SOUNDCLIP *AudioChannelsLock::SetChannel(int index, SOUNDCLIP* ch)
+{
+    _channels[index] = ch;
```

Comment:
Is it worth printing a debug warning here if we're overwriting a channel pointer? The channel would not get cleaned up otherwise. 

Another check might be to ensure we're not reusing a channel pointer, which I _think_ we're careful about but we'd have to check the cross fading code (that moves soundclips in/out of the music channel)

I'm fairly sure we're careful about freeing the channel first but it would be a good sanity check.

Reply:
I can add warnings, but this situation is complicated, because clips could be saved elsewhere and not deleted.
I do not see any good solution here except for storing SOUNDCLIP in a smart pointer (either unique_ptr or shared_ptr, depending on what kind of manipulations we want to allow).